### PR TITLE
Nh namespacing

### PR DESCRIPTION
**Pull request by nhunzaker**
This PR should allow for the deep namespacing of routes. For example, the following are now possible:

```javascript
var API = Gangway()

API.resource('users', {}, function(users) {
    users.resource('posts', {}, function (posts) {
        posts.route({
            wordCount: {
                path: 'word-count'
            }
        })
    })
})

console.log(API) // => { users: { create, read, update, delete, posts: { create, read, update, delete, wordCount } } }
```

Additionally:

```javascript
var API = Gangway()

var users = API.namespace('users').route({
    posts: {
        read() {
            path: 'posts/:id'
        }
    }
})

console.log(API) // => { users: { posts: { read } }
```

Nested parameter names for nested routes **are singularized**, following the convention of Rails. For example:

```javascript
var API = Gangway()

API.resource('users', {}, function(users) {
    users.resource('posts')
})
```

This will produce the following routes:

- `API.users.create()` at `/users`
- `API.users.read()` at `/users/{id?}`
- `API.users.update()` at `/users/{id}`
- `API.users.destroy()` at `/users/{id}`

and

- `API.users.posts.create()` at `/users/{user_id}/posts`
- `API.users.posts.read()` at `/users/{user_id}/posts/{id?}`
- `API.users.posts.update()` at `/users/{user_id}/posts/{id}`
- `API.users.posts.destroy()` at `/users/{user_id}/posts/{id}`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "url": "git://github.com/vigetlabs/gangway.git"
   },
   "dependencies": {
+    "inflected": "1.1.6",
```
Comment (nhunzaker) [inline package.json]: I want this to be hyper specific. Changes to inflection should not be taken lightly :P

File: src/api.js [hunk 2]
```diff
@@ -1,39 +1,50 @@
...
+API.prototype = {
+  ajax: ajax,
 
-  var API = {
-    config: config,
-    ajax: ajax,
+  namespace: function (key) {
+    var child = Object.create(this, {
+      segments: { value: this.segments.concat(key) }
+    })
```
Comment (nhunzaker) [inline src/api.js]: This creates a clone if the current instance of Gangway, overriding the `segments` property to include the new namespace.

**Comment by mackermedia** [comment]
Should
```
var users = API.namespace('users').route({
    posts: {
        read() {
            path: 'users/:id'
        }
    }
})
```
be
```
var users = API.namespace('users').route({
    posts: {
        read() {
            path: 'users/:user_id/posts'
        }
    }
})
```
?
That path thing reads funny to me as is.

File: src/api.js [hunk 3]
```diff
@@ -1,39 +1,50 @@
...
 
-    toString: function () {
-      return config.baseURL
-    },
+    this[key] = key in this ? assign(child, this[key]) : child
 
-    resolve: function (path, params) {
-      return url(config.baseURL, path, params)
-    },
+    return child
```
Comment (mackermedia) [inline src/api.js]: `child` is a funny name to me here. That or I'm not understanding what's going on in this method.

**Comment by nhunzaker** [comment]
@mackermedia I had that path wrong with way. It should have been `/posts/:id`.

File: src/api.js (hunk 3 above)
Comment (nhunzaker) [inline src/api.js]: This method creates a clone of the instance of Gangway that operates at a subpath. Everything else is shared. We want the namespaced configuration to inherit as much as possible. 

This is, basically, like creating another instance of Gangway with a different baseURL.

Comment (nhunzaker) [inline src/api.js]: Actually that might be a better way to do this... hmm.

Comment (mackermedia) [inline src/api.js]: Oh interesting...

**Comment by mackermedia** [comment]
I'm not sure I understand all of the code updates, but the tests look good :+1:

File: src/api.js (hunk 3 above)
Comment (nhunzaker) [inline src/api.js]: Alternative names... `branch`?

Comment (nhunzaker) [inline src/api.js]: `subset`?

**Comment by nhunzaker** [comment]
@mackermedia `path` _is_ awkward. The best alternative that I can think of would be a breaking update (now that we have namespaces) to convert the following:

```javascript
API.route({
    posts: {
        read() {
            path: 'posts/:id'
        }
    }
})
```

into

```javascript
API.namespace('posts').route({
    read() {
        path: ':id'
    }
})
```

Additionally, we could allow absolute paths to operate against the base URL like:

```
API.namespace('posts').route({
    read() {
        path: '/posts/:id'
    }
})
```

I think this feels more intuitive to how relative paths work in the browser. We'd definitely have to ship a `2.0` for this though, and it would break every project using Gangway. Still, the upgrade path would be pretty clear.

**Comment by nhunzaker** [comment]
Lastly, is `namespace` the best name for this?

File: src/assign.js [hunk 4]
```diff
@@ -10,5 +11,5 @@ module.exports = function assign (/* objects list */) {
     }
 
     return memo
-  }, {})
+  }, initial || {})
```
Comment (nhunzaker) [inline src/assign.js]: I needed to be able to merge properties into an object, which resulted in a updates to how `assign` was used in a couple of places. I may actually replace this with the `object-assign` NPM package. I'm evaluating browser support for some of the Object methods it uses.

File: src/api.js [hunk 5]
```diff
@@ -1,39 +1,54 @@
...
-var route = require('./route')
-var resource = require('./resource')
-var url = require('./url')
+var ajax      = require('./ajax')
+var assign    = require('./assign')
+var prepare   = require('./prepare')
+var resource  = require('./resource')
+var route     = require('./route')
+var url       = require('./url')
+var segmentize = require('./segmentize')
```
Comment (cwmanning) [inline src/api.js]: align `=`?

**Comment by cwmanning** [comment]
:+1:

File: test/namespace.test.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+var assert = require('assert')
+
+describe('Namespace', function() {
+
+  it ('can namespace a route', function() {
+    var api = new API()
+    var foo = api.namespace('foo')
+    var bar = foo.namespace('bar')
+
+    assert.deepEqual(bar.segments, [ 'foo', 'bar' ])
```
Comment (greypants) [inline test/namespace.test.js]: I might rename `foo` and `bar` to `parent` and `child` or `base` and `nested`. It wasn't clear at first that this nesting namespaces.

File: test/namespace.test.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    assert.equal(bar.toString(), '/foo/{foo_id}/bar')
+  })
+
+  it ('toString singularizes parameters segments', function() {
+    var api = new API()
+    var users = api.namespace('users')
+    var posts = users.namespace('posts')
+
+    assert.equal(posts.toString(), '/users/{user_id}/posts')
```
Comment (greypants) [inline test/namespace.test.js]: should `toString()` include the trailing `{id?}`?

File: test/namespace.test.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+  context('when a resource is created', function() {
+    var api   = new API()
+    var users = api.resource('users').namespace('users')
+
+    context('and a resource is generated', function() {
+      var notes = users.resource('notes')
+
+      it ('adds the resource to the parent namespace', function() {
+        assert.equal(api.users.notes.read, users.notes.read)
```
Comment (greypants) [inline test/namespace.test.js]: :tada:

**Comment by greypants** [comment]
:+1: this is great!

File: test/namespace.test.js (hunk 6 above)
Comment (nhunzaker) [inline test/namespace.test.js]: Good call. I was lazy :(

File: test/namespace.test.js (hunk 7 above)
Comment (nhunzaker) [inline test/namespace.test.js]: Good question. I don't think so. You absolutely need an `id` parameter for `update` and `delete`.

Comment (greypants) [inline test/namespace.test.js]: I'm sorry, I meant `/{id}`

Comment (nhunzaker) [inline test/namespace.test.js]: Ah sorry. Still no. `posts.toString()` is going to return the base path name for user posts. Routes build on this, adding the `{id}` when necessary.

**Pull request merged**